### ISSUE TITLE
Docker issues with plugins

### ISSUE THREAD

**Issue by andreialecu** [Countly/countly-server#869]
I'm running Countly on docker containers inside Amazon ECS and hosting the database on MongoDB Atlas.

With this setup, the Countly containers are assigned a minimum of 400MB of RAM, with a hard cap of 800MB. There are no problems with regular usage with this setup, since the heavy lifting is on the mongo side, and this is a relatively low traffic installation.

However, after configuring anything plugin related, there seems that the container hangs after logging `Preparing production files`.

```
2019-03-23 18:39:46 Preparing production files
2019-03-23 18:39:46 Done uninstalling plugin "populator"
2019-03-23 18:39:46 Uninstalling plugin "populator"...
```

I assume this is because of the limited resources dedicated to the container. I have tried disabling `Production mode` in the configuration via the dashboard, but I'm not sure where that specific flag is persisted, or if it is persisted at all, because it doesn't seem to do anything. 

CPU usage goes to 100% and it doesn't seem to ever finish unless rebooting the container. Docker itself is unresponsive and cannot even stop the task.

I can only get Countly back running by changing the plugin manually in the `plugins` collection back to being installed.

So basically the problem, or question here is. Why does disabling 'Production mode' in the dashboard not disable the 'Preparing production files' step? And/or why does simply disabling a plugin result in a preparing step that takes 100% CPU on a low resource VM?

**Comment by andreialecu** [comment]
In the `plugins` collection, it looks like `frontend` has `production: false` in it.

**Comment by ar2rsawseen** [comment]
Hello,
all I can say, is countly requires minimum 2Gb RAM to run.
We cannot guarantee anything underneath that.

Yes, it probably can be tweaked to run under less, by disabling some features, etc.
But honestly no one has much time for that, because that is not what most of the clients want.

But if you are up for it, we are open for PR on any improvements and everything is negotiable :)

The decision under your question is simple. Preparing production files takes quite some load and time.
If we don't do that on plugin change (even if production mode is disabled), it means we will need to do that when production mode gets enabled. And we don't want to make user wait 20 seconds on runtime configuration changes.

Additionally, it is possible to disable production mode on server level, but enable on user level that accesses the dashboard, and then we would need to regenerate files every time someone changes their user configuration.

Plugin enabling/disabling, on the other hand, is meant to restart nodejs process, so it takes lots of time either way, and we are changing to async check on browser side now to anticipate even longer loads now.

So as a result we regenerate production files when there are actually changes, and not when config is set. 
This seems to be both efficient on UX and performance side :)

**Comment by andreialecu** [comment]
I know 2GB is the recommended minimum, but that is with MongoDB on the same server as Countly, no - which is supposed to take most of the load? 

I've noticed that under real usage, Countly is very happy with a couple of hundred MBs of memory for itself.

This could potentially be a different problem, perhaps related to the Docker container, because I can see `Preparing production files` logged, from https://github.com/Countly/countly-server/blob/fa02e34320dd17ae5c72da3073a0de0aa663a1a0/plugins/pluginManager.js#L794 but the following log line should be this: https://github.com/Countly/countly-server/blob/fa02e34320dd17ae5c72da3073a0de0aa663a1a0/Gruntfile.js#L196 which I'm not seeing at all. It seems to hang before that.

**Comment by andreialecu** [comment]
FWIW, I'm using this, unmodified: https://github.com/Countly/countly-server/blob/master/Dockerfile-core

**Comment by andreialecu** [comment]
Alright, there's no log from Grunt. I've run the docker file locally and I can see this logged:

```
Uninstalling plugin "populator"...
Done uninstalling plugin "populator"
Preparing production files
Done preparing production files with null / "Running \"plugins\" task\nPreparing production files for following plugins: [\"mobile\",\"web\",\"desktop\",\"plugins\",\"density\",\"locale\",\"browser\",\"sources\",\"views\",\"enterpriseinfo\",\"logger\",\"systemlogs\",\"errorlogs\",\"reports\",\"crashes\",\"push\",\"star-rating\",\"slipping-away-users\",\"compare\",\"server-stats\",\"dbviewer\",\"assistant\",\"plugin-upload\",\"times-of-day\",\"compliance-hub\",\"video-intelligence-monetization\",\"alerts\",\"onboarding\"]\nDone preparing production files\n\nRunning \"concat:plugins\" (concat) task\n\nRunning \"uglify:plugins\" (uglify) task\n>> 1 file created 2.52 MB → 1.35 MB\n\nRunning \"copy:plugins\" (copy) task\nCopied 60 files\n\nRunning \"cssmin:plugins\" (cssmin) task\n>> 1 file created. 209.74 kB → 167.5 kB\n\nRunning \"locales\" task\nPreparing locale files for core & plugins: [\"mobile\",\"web\",\"desktop\",\"plugins\",\"density\",\"locale\",\"browser\",\"sources\",\"views\",\"enterpriseinfo\",\"logger\",\"systemlogs\",\"errorlogs\",\"reports\",\"crashes\",\"push\",\"star-rating\",\"slipping-away-users\",\"compare\",\"server-stats\",\"dbviewer\",\"assistant\",\"plugin-upload\",\"times-of-day\",\"compliance-hub\",\"video-intelligence-monetization\",\"alerts\",\"onboarding\"]\nDone preparing locale files\n\nRunning \"concat:locales_\" (concat) task\n\nRunning \"concat:locales_ar\" (concat) task\n\nRunning \"concat:locales_bs\" (concat) task\n\nRunning \"concat:locales_ca\" (concat) task\n\nRunning \"concat:locales_de\" (concat) task\n\nRunning \"concat:locales_el\" (concat) task\n\nRunning \"concat:locales_es\" (concat) task\n\nRunning \"concat:locales_et\" (concat) task\n\nRunning \"concat:locales_fa\" (concat) task\n\nRunning \"concat:locales_fr\" (concat) task\n\nRunning \"concat:locales_hi\" (concat) task\n\nRunning \"concat:locales_hu\" (concat) task\n\nRunning \"concat:locales_it\" (concat) task\n\nRunning \"concat:locales_ja\" (concat) task\n\nRunning \"concat:locales_ko\" (concat) task\n\nRunning \"concat:locales_lv\" (concat) task\n\nRunning \"concat:locales_my\" (concat) task\n\nRunning \"concat:locales_nb\" (concat) task\n\nRunning \"concat:locales_nl\" (concat) task\n\nRunning \"concat:locales_pt\" (concat) task\n\nRunning \"concat:locales_ro\" (concat) task\n\nRunning \"concat:locales_ru\" (concat) task\n\nRunning \"concat:locales_sl\" (concat) task\n\nRunning \"concat:locales_sv\" (concat) task\n\nRunning \"concat:locales_tr\" (concat) task\n\nRunning \"concat:locales_vi\" (concat) task\n\nRunning \"concat:locales_zh\" (concat) task\n\nRunning \"concat:locales_pl\" (concat) task\n\nDone.\n"
Restarting Countly ...
Done restarting countly with {"killed":false,"code":1,"signal":null,"cmd":"sudo countly restart"} / "sudo: no tty present and no askpass program specified\n" / ""
error: {"killed":false,"code":1,"signal":null,"cmd":"sudo countly restart"}
stderr: "sudo: no tty present and no askpass program specified\n"
Uninstalling plugin "populator"...
Done uninstalling plugin "populator"
Preparing production files
Done preparing production files with null / "Running \"plugins\" task\nPreparing production files for following plugins: [\"mobile\",\"web\",\"desktop\",\"plugins\",\"density\",\"locale\",\"browser\",\"sources\",\"views\",\"enterpriseinfo\",\"logger\",\"systemlogs\",\"errorlogs\",\"reports\",\"crashes\",\"push\",\"star-rating\",\"slipping-away-users\",\"compare\",\"server-stats\",\"dbviewer\",\"assistant\",\"plugin-upload\",\"times-of-day\",\"compliance-hub\",\"video-intelligence-monetization\",\"alerts\",\"onboarding\"]\nDone preparing production files\n\nRunning \"concat:plugins\" (concat) task\n\nRunning \"uglify:plugins\" (uglify) task\n>> 1 file created 2.52 MB → 1.35 MB\n\nRunning \"copy:plugins\" (copy) task\nCopied 60 files\n\nRunning \"cssmin:plugins\" (cssmin) task\n>> 1 file created. 209.74 kB → 167.5 kB\n\nRunning \"locales\" task\nPreparing locale files for core & plugins: [\"mobile\",\"web\",\"desktop\",\"plugins\",\"density\",\"locale\",\"browser\",\"sources\",\"views\",\"enterpriseinfo\",\"logger\",\"systemlogs\",\"errorlogs\",\"reports\",\"crashes\",\"push\",\"star-rating\",\"slipping-away-users\",\"compare\",\"server-stats\",\"dbviewer\",\"assistant\",\"plugin-upload\",\"times-of-day\",\"compliance-hub\",\"video-intelligence-monetization\",\"alerts\",\"onboarding\"]\nDone preparing locale files\n\nRunning \"concat:locales_\" (concat) task\n\nRunning \"concat:locales_ar\" (concat) task\n\nRunning \"concat:locales_bs\" (concat) task\n\nRunning \"concat:locales_ca\" (concat) task\n\nRunning \"concat:locales_de\" (concat) task\n\nRunning \"concat:locales_el\" (concat) task\n\nRunning \"concat:locales_es\" (concat) task\n\nRunning \"concat:locales_et\" (concat) task\n\nRunning \"concat:locales_fa\" (concat) task\n\nRunning \"concat:locales_fr\" (concat) task\n\nRunning \"concat:locales_hi\" (concat) task\n\nRunning \"concat:locales_hu\" (concat) task\n\nRunning \"concat:locales_it\" (concat) task\n\nRunning \"concat:locales_ja\" (concat) task\n\nRunning \"concat:locales_ko\" (concat) task\n\nRunning \"concat:locales_lv\" (concat) task\n\nRunning \"concat:locales_my\" (concat) task\n\nRunning \"concat:locales_nb\" (concat) task\n\nRunning \"concat:locales_nl\" (concat) task\n\nRunning \"concat:locales_pt\" (concat) task\n\nRunning \"concat:locales_ro\" (concat) task\n\nRunning \"concat:locales_ru\" (concat) task\n\nRunning \"concat:locales_sl\" (concat) task\n\nRunning \"concat:locales_sv\" (concat) task\n\nRunning \"concat:locales_tr\" (concat) task\n\nRunning \"concat:locales_vi\" (concat) task\n\nRunning \"concat:locales_zh\" (concat) task\n\nRunning \"concat:locales_pl\" (concat) task\n\nDone.\n"
Restarting Countly ...
Done restarting countly with {"killed":false,"code":1,"signal":null,"cmd":"sudo countly restart"} / "sudo: no tty present and no askpass program specified\n" / ""
error: {"killed":false,"code":1,"signal":null,"cmd":"sudo countly restart"}
stderr: "sudo: no tty present and no askpass program specified\n"
```

The plugin however doesn't seem to uninstall properly. it is unchecked in the plugins list but I still see the Populator link in the sidebar and it is still installed. This leaves countly in some sort of bad state, because I cannot seem to be able to reenable the plugin from the plugins manager. The "Save Changes" button never activates.

I assume more work needs to be done for full docker compatibility.

**Comment by andreialecu** [comment]
The plugin architecture right now seems like it is not really compatible with clustering. It doesn't seem like a good idea for every node in the cluster to have to rebuild production files for plugin changes. 

This will be an issue with Docker, by default, because the file system is ephemeral, and any plugin change from the default of when the Countly docker image was created will mean that Docker containers will need to rebuild plugins on every single start of the container.

Perhaps the 'production' files could be pre-built and distributed with plugins, or they could be stored in the database and served from there (gridfs?). Seems like a `plugins.json` file is also read from the file system of each node, which doesn't make sense in a clustered scenario.

I don't know enough about Countly innards but I could potentially try my hand at a PR to move this functionality to the Countly gridfs implementation. What do you think?

**Comment by ar2rsawseen** [comment]
It is cluster compatible, as all changes are run on master process, not cluster workers.
In multi server environment it is run once on every server.

You still see a link because production files are not modified. I guess if you dont have production mode enabled and load separate files in dashboard it should be fine. But might need clearing cache

But in most cases, this is not what you would do on deployment. Ideally you would modify docker image on separate machine and redeploy updated docker image across your deployment.

Or even install Countly separately, make all the needed configurations with plugins, etc and create your own docker image.

Because due to ephemeral file system, lots of things won't work, as you say. So Countly is not meant to be configured in docker container, but rather creating a docker image with configurations already set in and using that.

**Comment by andreialecu** [comment]
Sounds good. Is there a way to disable the Plugin Manager in the dashboard then, since this might lead to mistakes? Is there a configuration setting for this?

Also, any clues on how to modify https://github.com/Countly/countly-server/blob/master/Dockerfile-core to modify the list of plugins installed by default?

**Comment by ar2rsawseen** [comment]
Initial list of plugins is in plugins/plugins.json so modifying it before installation would ensure only those plugins are installed when installing Countly.

About hiding plugin manager, not really possible out of the box, but could be done with injected theme or separate plugin to do any other UI modifications, etc as you see fit.

**Comment by ar2rsawseen** [comment]
Also I think it was possible to use original docker image, then made changes to file system needed and do `docker commit` to create your own image with the changes

But basically we never experienced before the hang on that side in docker container, but we also never ran it on such low level RAM.
We go with minimal calculation of 10M datapoints per month, which requires separate 2Gb for nodejs and 4Gb for MongoDB
And we are not testing below those limits:
https://code.count.ly/helpers.html

**Comment by andreialecu** [comment]
Thanks for the tip regarding plugins.json.

You could still run the 10M datapoints scenario on a cluster of 2+ x 1GB nodes. (5M each). This would bring more scaling benefits and is potentially cheaper than running a single higher memory node.

Just a thought, but it might be a good idea to just disable Plugin Manager entirely for docker installations since it seems like it is not really working right.

There is an INSIDE_DOCKER env variable which could be used. Even hiding the Plugins link from the Management menu would help. I've killed Countly several times before figuring out what was wrong.

**Comment by ar2rsawseen** [comment]
Yes, but then again, you won't be able to launch original container, do modifications in plugins, commit and create your own container, if you dont have UI for modifications :)

Either way, leaving it up for @iartem as he is the one managing docker side :)

**Comment by andreialecu** [comment]
Could be a different env variable, INSIDE_DOCKER_ENABLE_PLUGIN_UI=1 - so you could bypass and get the UI.

**Comment by iartem** [comment]
@andreialecu we've had an issue with docker plugin modifications, it's fixed in [this commit](https://github.com/Countly/countly-server/commit/fa02e34320dd17ae5c72da3073a0de0aa663a1a0), yet it's not in a tagged release yet. I guess all your issues are more or less coming from this issue. In case you need it fixed now, you can apply that fix directly to your image.

**Comment by andreialecu** [comment]
@iartem I have seen the commit, but that only partially helps. If the docker container dies, the plugins need to be re-uglified on every start. That is an expensive process, and can create problems, especially on cloud providers - the process may only finish after an extensive period of time, and the host will most likely mark the container as unhealthy because of failing health checks, killing it, thus creating an endless loop where the process never gets to start properly, and consumes all host resources over and over.

I have had Countly kill full Container instances because of this. It affected other docker containers on the same host because it would take up all the resources building and rebuilding plugins. I had to create its own separate docker cluster with dedicated machines to ensure this doesn't happen again.

It's easy to make a mistake using the Plugin Manager and modify something and get stuck in a bad state. There should at least be a warning when INSIDE_DOCKER is detected that changing Plugins is a bad idea, and instead the docker image should be prebuilt with proper plugins and redeployed.

Like I mentioned in a previous post, even though 2GB is recommended, I still think a cluster of 2+ 1GB or lower instances would work in the same as a single higher memory single instance. And I have verified this in production, memory usage is way below even 400MB on a production Countly instance.

Also, it's somewhat of a good practice for all stateless type apps, which Countly seems to be a part of, to not consume a ton of memory in the first place. Scaling horizontally is always preferred to scaling vertically, in a cloud-first world.

**Comment by iartem** [comment]
@andreialecu I'll start from the end. As people say... Scaling horizontally is overrated :) In Countly case, having 2 separate containers each on 1 core will **always** be worse in terms of CPU, memory & storage - all of those. Simply because of how Countly works, and it works in a standard nodejs way - using `cluster` module to spread the load across multiple cores. A set of 2 1GB 1-core instances will **always** work worse than 1 2GB 2-core instance. This is true until the moment one core is incapable of handling all incoming connections, and to our experience, it's around 8 cores or so. Only after that moment it's more efficient to have multiple servers than just one.

Now all your problems seem to come from the fact that your cloud provider marks an instance as unhealthy after a very short period of time. So why not to increase that time? Your suggestion is far more than you think as what you actually suggest is getting all the cluster consistency checks we have out of countly and put it solely on k8s or devops. That would produce way more problems than just increasing the health check interval or number of checks needed.

In case you cannot do it for some reason, I'd suggest replacing [this](https://github.com/Countly/countly-server/blob/master/plugins/pluginManager.js#L795-L805) with just `callback()` and testing it well.

**Comment by andreialecu** [comment]
@iartem It's not just the unhealthy timeout. It never actually finishes "Preparing production files". It marks it unhealthy, but the docker container stays stuck and cannot even be killed with CPU usage at 100%. It eats up the whole resources of the specific host VM and nothing is responsive any more. I have waited several hours for it to clear up and it didn't.

I'm not sure why this happens, must be some extensive swapping going on.

Regarding scaling horizontally, I respectully disagree. :) With horizontal scaling you can assign very low memory to instances and set up auto-scaling rules, so everything scales elastically and costs and resource usage is kept to absolute minimum levels. Availability is much higher and any traffic surge is handled perfectly.

Most apps will have uneven loads, so a lot of the time having a high memory instance assigned will mean a lot of the resources are wasted.

Serverless is becoming the new trend for a reason. :)

**Comment by iartem** [comment]
Well, I tried :) Last comments on the topic:

1. With any substantial load, memory usage will spike to 3-4GB per core. You cannot "assign" 1GB and think it won't swap or crash. Again, 4 x 1GB/1-core instances for sure are less (in reqs/sec) than 1 x 4GB/4-core instance and most probably even less than 1 x 4GB/1-core.
2. Everything from (1) is because of a simple reason: with default settings, we maximize the performance we can squeeze out of one CPU core. To keep it in 1GB area you'll have to have a ridiculous maximum number of mongo connections... No idea how much, but it should be around 50 I guess, probably even less. Meaning CPU will be either underloaded for this memory usage, or you'll run multiple containers effectively on a single core. Which is always less effective than 1 process per CPU core and brings us back to (1).
3. Also respectfully, but I have to point out :) `everything scales elastically` & `resource usage is kept to absolute minimum levels` nearly never happens in the real world unless you have tons of microservices. Countly is not a microservice. You'll have to pay several hundreds of MB of RAM per container for an ability to scale elastically in case of Countly. And this extra is not an `absolute minimum levels`.


Grunt / uglify must use quite a lot of CPU, that's true. But I've never seen this behavior before. Again, Countly is open source, just turn production files generation off by commenting out that line.

**Comment by andreialecu** [comment]
I have opened this to point out a valid issue. 

**With a docker container, it's an antipattern to have to rebuild production files and then save them locally, on every container start.**  There should be at least a warning that they should be baked into the image itself. With any wide adoption of Countly, this is an issue that potentially will affect potential customers, such as myself. I have almost given up several times due to the stability issues.

Regarding your comments above, I would certainly like to understand why Countly is a memory hog like this. I haven't verified that it is, but all documentation points out this way. Actual APIs we're using in production get by MUCH more efficiently that the load numbers claimed by Countly in this case. One would assume most of the work is done by MongoDB, which is separate anyway so doesn't interfere at all. Also, docker containers can spin up really fast, and take load almost instantly, so in practice in the real world, scaling horizontally actually works magically.

But I do realize by looking at the way setup works that Countly was started before this time, there's technical debt to be paid (was interesting seeing `grunt` after all this time), and probably a dev with a fresh overview of modern clustering techniques needs to take a look and reassess. 

Luckily however, it still seems usable, with some minor quirks. At least a documentation warning would be good, for the Docker/Preparing production files scenario.
